argoproj/argo
# feat(ui): Make UI errors recoverable. Fixes #3666

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] ~~I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.~~ N/A
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

**I recommend the reviewer and I sit down and go over the changes.**

# Goals

The goal of this PR is to make the UI usable under error scenarios. Typically if an error occurs, we throw it and a global `<ErrorBoundary/>` render a generic "ops!" page. The user must then reload. This looses the user's context as well as any data entered into forms - very annoying.

Errors are now more localised in the interface. E.g. an error listing workflows just shows the error and keeps the page usable. The user can then change their search parameters. If the error was a 403, this would most useful.

# Changes

* Refactored handling of managed namespace in UI. Rather than have each component enforce this, it is done in `<App/>`. This simplifies many component.
* Add automatic redirection to the correct URLs for managed namespaces.
* Add `malformed-*.yaml` manifests to aid testing with malformed manifests.
* Removed `throw this.state.error` and replaced with `<ErrorNotice/>` so that the UI remains usable on error.
* `<ResourceEditor/>` now displays submission error in a way so that you can edit and retry.
* Simplified history by adding `saveHistory` functions and `this.url =`.

# Testing Notes

**I'm not expecting the reviewers to do this - this is for my notes**.

Does the UI stay usable?

* Try listing and deep-linking malformed resources. UI should still be usable.
* List each type of resource. 
* Creating each type of resource (both valid and malformed). Check namespace is correct.
* Deleting each type of resource. Check namespace is correct.
* Submitting each type of submittable resources. Check the correct namespace is used.

Can we deal with really bad errors?

* Run `yarn start` and turn off Argo Server so that you get gateway errors, try all pages.

Does login still work?

* Run in `make AUTH_MODE=client start` and verify that you are redirect to login page.

Does namespace preference still work?

* Run in `make NAMESPACED=false start` clear local storage and verify no redirection occurs (you should just list the at cluster scope).
* Change namespace in each page and verify that your local preference (stored in local storage) is updated.

# Known Issues

* The namespace preference does not affect URLs unless you reload the page. Fix to this is complex vs benefit IMHO.
